testing.rst

```diff
@@ -11,377 +11,513 @@ using both functional and unit tests.
+            xsi:schemaLocation="http://symfony.com/schema/dic/services
+                https://symfony.com/schema/dic/services/services-1.0.xsd
+                http://symfony.com/schema/dic/twig
+                https://symfony.com/schema/dic/twig/twig-1.0.xsd">
 
-Functional Tests
-----------------
+            <framework:config strict-variables="true"/>
+        </container>
 
-Functional tests check the integration of the different layers of an
-application (from the routing to the views). They are no different from unit
-tests as far as PHPUnit is concerned, but they have a very specific workflow:
+    .. code-block:: php
 
-* Make a request;
-* Click on a link or submit a form;
-* Test the response;
-* Rinse and repeat.
+        // config/packages/test/twig.php
+        $container->loadFromExtension('twig', [
+            'strict_variables' => true,
+        ]);
 
-Before creating your first test, install these packages that provide some of the
-utilities used in the functional tests:
+You can also use a different environment entirely, or override the default
+debug mode (``true``) by passing each as options to the ``bootKernel()``
+method::
 
-.. code-block:: terminal
+    self::bootKernel([
+        'environment' => 'my_test_env',
+        'debug'       => false,
+    ]);
 
-    $ composer require --dev symfony/browser-kit symfony/css-selector
+.. tip::
 
-Your First Functional Test
-~~~~~~~~~~~~~~~~~~~~~~~~~~
-Functional tests are PHP files that typically live in the ``tests/Controller``
-directory of your application. If you want to test the pages handled by your
-``PostController`` class, start by creating a new ``PostControllerTest.php``
-file that extends a special ``WebTestCase`` class.
+    It is recommended to run your test with ``debug`` set to ``false`` on
+    your CI server, as it significantly improves test performance.
 
-As an example, a test could look like this::
+Customizing Environment Variables
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-    // tests/Controller/PostControllerTest.php
-    namespace App\Tests\Controller;
+If you need to customize some environment variables for your tests (e.g. the
+``DATABASE_URL`` used by Doctrine), you can do that by overriding anything you
+need in your ``.env.test`` file:
 
-    use Symfony\Bundle\FrameworkBundle\Test\WebTestCase;
+.. code-block:: text
 
-    class PostControllerTest extends WebTestCase
+    # .env.test
+
+    # ...
+    DATABASE_URL="mysql://db_user:db_password@127.0.0.1:3306/db_name_test?serverVersion=5.7"
+
+In the test environment, these env files are read (if vars are duplicated
+in them, files lower in the list override previous items):
+
+#. ``.env``: containing env vars with application defaults;
+#. ``.env.test``: overriding/setting specific test values or vars;
+#. ``.env.test.local``: overriding settings specific for this machine.
+
+.. caution::
+
+    The ``.env.local`` file is **not** used in production, to make each
```

Comment:
```suggestion
    The ``.env.local`` file is **not** used in the test environment, to make each
```

Reply:
This change is marked as "resolved".  I think it was a typo, shouldn't this be corrected?